File: src/csharp/Foundgine.Core/Semantic/Planning/RewriteRuleComposer.cs
namespace Foundgine.Core.Semantic.Planning;

/// <summary>
/// Deterministically composes rewrite rules while enforcing ordering, conflicts,
/// selection, idempotence and termination limits. Proof obligations remain per application.
/// </summary>
public sealed class RewriteRuleComposer
{
    private readonly IReadOnlyList<IPlanRewriteRule> _orderedRules;
    private readonly RewriteRuleCompositionOptions _options;
    private readonly RewriteRuleSelector _selector;
    private readonly IProviderCostEstimator? _providerCostEstimator;

    public RewriteRuleComposer(
        IEnumerable<IPlanRewriteRule> rules,
        RewriteRuleCompositionOptions? options = null)
    {
        ArgumentNullException.ThrowIfNull(rules);
        _options = (options ?? new()).Validate();
        _orderedRules = OrderRules(rules.ToArray());
        _providerCostEstimator = _options.ProviderCostEstimator;
        _selector = new RewriteRuleSelector(
            _options.SelectionPolicy,
            _providerCostEstimator,
            _options.ProviderCostSelectionPolicy);
    }

    public RewriteRuleCompositionResult Compose(SemanticPlan plan)
    {
        ArgumentNullException.ThrowIfNull(plan);

        var current = plan;
        var applications = new List<PlanRewriteRuleResult>();
        var selectionHistory = new List<RewriteRuleCandidate>();
        var providerSelectionHistory = new List<ProviderAwareRewriteRuleCandidate>();
        var seenPlans = new HashSet<string>(StringComparer.Ordinal)
        {
            SemanticPlanFingerprint.Create(current)
        };
        var appliedIdempotent = new HashSet<string>(StringComparer.Ordinal);
        var appliedRules = new HashSet<string>(StringComparer.Ordinal);

        for (var pass = 0;; pass++)
        {
            var eligible = _orderedRules
                .Where(rule => !(rule.IsIdempotent && appliedIdempotent.Contains(rule.Name)))
                .Where(rule => rule.MustRunAfter.All(appliedRules.Contains))
                // A rule that must run before another rule is not itself
                // blocked. Instead, block the dependent rule while the
                // prerequisite remains applicable. This is important when
                // selection uses benefit/cost rather than raw priority.
                .Where(rule => _orderedRules
                    .Where(prerequisite => prerequisite.MustRunBefore.Contains(rule.Name, StringComparer.Ordinal))
                    .All(prerequisite =>
                        appliedRules.Contains(prerequisite.Name) ||
                        !prerequisite.CanApply(current)))
                .Where(rule => !rule.ConflictsWith.Any(appliedRules.Contains))
                .ToArray();

            string? selectedRuleName;
            if (_providerCostEstimator is not null)
            {
                var selectedProvider = _selector.SelectProviderAware(current, eligible);
                if (selectedProvider is null)
                {
                    return new RewriteRuleCompositionResult(
                        current,
                        applications,
                        applications.Sum(x => x.CostImpact),
                        true,
                        selectionHistory,
                        providerSelectionHistory);
                }

                providerSelectionHistory.Add(selectedProvider);
                selectedRuleName = selectedProvider.RuleName;
            }
            else
            {
                var selected = _selector.Select(current, eligible);
                if (selected is null)
                {
                    return new RewriteRuleCompositionResult(
                        current,
                        applications,
                        applications.Sum(x => x.CostImpact),
                        true,
                        selectionHistory,
                        providerSelectionHistory);
                }

                selectionHistory.Add(selected);
                selectedRuleName = selected.RuleName;
            }

            var rule = _orderedRules.First(r => StringComparer.Ordinal.Equals(r.Name, selectedRuleName));

            if (applications.Count >= _options.MaxRuleApplications)
                throw new InvalidOperationException(
                    "Rewrite rule composition exceeded the maximum rule-application budget.");

            var candidate = rule.Apply(current);
            if (ReferenceEquals(candidate, current))
            {
                appliedRules.Add(rule.Name);
                if (rule.IsIdempotent)
                    appliedIdempotent.Add(rule.Name);
                continue;
            }

            var result = SemanticPlanOptimizer.ApplyRule(rule, current, candidate);
            if (!result.IsSatisfied)
                throw new InvalidOperationException(
                    $"Rewrite rule '{rule.Name}' did not satisfy its proof obligations.");

            var currentFingerprint = SemanticPlanFingerprint.Create(current);
            var fingerprint = SemanticPlanFingerprint.Create(candidate);

            // A rule may rebuild an equivalent object graph while producing
            // exactly the same executable plan fingerprint. That is a
            // canonicalization no-op, not a rewrite cycle. Treat it as
            // unchanged and allow the composer to mark the rule complete.
            if (StringComparer.Ordinal.Equals(currentFingerprint, fingerprint))
            {
                appliedRules.Add(rule.Name);
                if (rule.IsIdempotent)
                    appliedIdempotent.Add(rule.Name);
                continue;
            }

            if (!seenPlans.Add(fingerprint))
                throw new InvalidOperationException(
                    $"Rewrite rule composition detected a cycle at rule '{rule.Name}'.");
            if (seenPlans.Count > _options.MaxPlanVisits)
                throw new InvalidOperationException("Rewrite rule composition exceeded the maximum plan-visit budget.");

            applications.Add(result);
            current = candidate;
            appliedRules.Add(rule.Name);
            if (rule.IsIdempotent)
                appliedIdempotent.Add(rule.Name);

            if (pass >= _options.MaxRuleApplications)
                throw new InvalidOperationException(
                    "Rewrite rule composition exceeded the maximum composition passes.");
        }
    }

    private static IReadOnlyList<IPlanRewriteRule> OrderRules(IReadOnlyList<IPlanRewriteRule> rules)
    {
        var byName = rules.ToDictionary(r => r.Name, StringComparer.Ordinal);
        foreach (var rule in rules)
        {
            foreach (var dependency in rule.MustRunAfter.Concat(rule.MustRunBefore))
            {
                if (!byName.ContainsKey(dependency))
                    throw new InvalidOperationException(
                        $"Rewrite rule '{rule.Name}' references unknown rule '{dependency}'.");
            }

            foreach (var conflict in rule.ConflictsWith)
            {
                if (byName.ContainsKey(conflict) &&
                    byName[conflict].ConflictsWith.Contains(rule.Name, StringComparer.Ordinal))
                    throw new InvalidOperationException(
                        $"Rewrite rules '{rule.Name}' and '{conflict}' conflict and cannot be composed.");
            }
        }

        var edges = rules.ToDictionary(r => r.Name, _ => new HashSet<string>(StringComparer.Ordinal),
            StringComparer.Ordinal);
        var indegree = rules.ToDictionary(r => r.Name, _ => 0, StringComparer.Ordinal);
        foreach (var rule in rules)
        {
            foreach (var after in rule.MustRunAfter)
                AddEdge(after, rule.Name);
            foreach (var before in rule.MustRunBefore)
                AddEdge(rule.Name, before);
        }

        var queue = new PriorityQueue<IPlanRewriteRule, (int Priority, string Name)>();
        foreach (var rule in rules.Where(r => indegree[r.Name] == 0))
            queue.Enqueue(rule, (rule.Priority, rule.Name));

        var ordered = new List<IPlanRewriteRule>(rules.Count);
        while (queue.TryDequeue(out var rule, out _))
        {
            ordered.Add(rule);
            foreach (var next in edges[rule.Name])
            {
                if (--indegree[next] == 0)
                {
                    var candidate = byName[next];
                    queue.Enqueue(candidate, (candidate.Priority, candidate.Name));
                }
            }
        }

        if (ordered.Count != rules.Count)
            throw new InvalidOperationException("Rewrite rule composition contains an ordering cycle.");

        return ordered;

        void AddEdge(string from, string to)
        {
            if (edges[from].Add(to))
                indegree[to]++;
        }
    }
}

File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/planning/RewriteRuleComposer.java
package com.foundgine.core.semantic.planning;

import java.util.*;

/**
 * Deterministically composes rewrite rules while enforcing ordering, conflicts
 * and budgets.
 */
public final class RewriteRuleComposer {
	private final List<IPlanRewriteRule> orderedRules;
	private final RewriteRuleCompositionOptions options;
	private final RewriteRuleSelector selector;

	public RewriteRuleComposer(Iterable<IPlanRewriteRule> rules) {
		this(rules, null);
	}

	public RewriteRuleComposer(Iterable<IPlanRewriteRule> rules, RewriteRuleCompositionOptions options) {
		Objects.requireNonNull(rules);
		this.options = (options == null ? new RewriteRuleCompositionOptions() : options).validate();
		this.orderedRules = orderRules(toList(rules));
		this.selector = new RewriteRuleSelector(this.options.selectionPolicy(), this.options.providerCostEstimator(),
				this.options.providerCostSelectionPolicy());
	}

	public RewriteRuleCompositionResult compose(SemanticPlan plan) {
		Objects.requireNonNull(plan);
		var current = plan;
		var applications = new ArrayList<PlanRewriteRuleResult>();
		var history = new ArrayList<RewriteRuleCandidate>();
		var providerHistory = new ArrayList<ProviderAwareRewriteRuleCandidate>();
		var seen = new HashSet<String>();
		seen.add(SemanticPlanFingerprint.create(current));
		var appliedIdempotent = new HashSet<String>();
		var appliedRules = new HashSet<String>();
		for (int pass = 0;; pass++) {
			var eligible = new ArrayList<IPlanRewriteRule>();
			for (var rule : orderedRules) {
				if (rule.isIdempotent() && appliedIdempotent.contains(rule.name()))
					continue;
				if (rule.mustRunAfter().stream().anyMatch(x -> !appliedRules.contains(x)))
					continue;
				boolean blocked = false;
				for (var prerequisite : orderedRules)
					if (prerequisite.mustRunBefore().contains(rule.name())
							&& !appliedRules.contains(prerequisite.name()) && prerequisite.canApply(current)) {
						blocked = true;
						break;
					}
				if (blocked)
					continue;
				if (rule.conflictsWith().stream().anyMatch(appliedRules::contains))
					continue;
				eligible.add(rule);
			}
			String selectedName;
			if (options.providerCostEstimator() != null) {
				var selected = selector.selectProviderAware(current, eligible);
				if (selected == null)
					return result(current, applications, history, providerHistory);
				providerHistory.add(selected);
				selectedName = selected.ruleName();
			} else {
				var selected = selector.select(current, eligible);
				if (selected == null)
					return result(current, applications, history, providerHistory);
				history.add(selected);
				selectedName = selected.ruleName();
			}
			IPlanRewriteRule rule = orderedRules.stream().filter(r -> r.name().equals(selectedName)).findFirst()
					.orElseThrow();
			if (applications.size() >= options.maxRuleApplications())
				throw new IllegalStateException(
						"Rewrite rule composition exceeded the maximum rule-application budget.");
			var candidate = rule.apply(current);
			if (candidate == current) {
				appliedRules.add(rule.name());
				if (rule.isIdempotent())
					appliedIdempotent.add(rule.name());
				continue;
			}
			var result = applyRule(rule, current, candidate);
			if (!result.isSatisfied())
				throw new IllegalStateException(
						"Rewrite rule '" + rule.name() + "' did not satisfy its proof obligations.");
			var beforeFingerprint = SemanticPlanFingerprint.create(current);
			var fingerprint = SemanticPlanFingerprint.create(candidate);
			if (beforeFingerprint.equals(fingerprint)) {
				appliedRules.add(rule.name());
				if (rule.isIdempotent())
					appliedIdempotent.add(rule.name());
				continue;
			}
			if (!seen.add(fingerprint))
				throw new IllegalStateException(
						"Rewrite rule composition detected a cycle at rule '" + rule.name() + "'.");
			if (seen.size() > options.maxPlanVisits())
				throw new IllegalStateException("Rewrite rule composition exceeded the maximum plan-visit budget.");
			applications.add(result);
			current = candidate;
			appliedRules.add(rule.name());
			if (rule.isIdempotent())
				appliedIdempotent.add(rule.name());
			if (pass >= options.maxRuleApplications())
				throw new IllegalStateException("Rewrite rule composition exceeded the maximum composition passes.");
		}
	}

	private static RewriteRuleCompositionResult result(SemanticPlan p, List<PlanRewriteRuleResult> a,
			List<RewriteRuleCandidate> h, List<ProviderAwareRewriteRuleCandidate> ph) {
		return new RewriteRuleCompositionResult(p, a, a.stream().mapToDouble(PlanRewriteRuleResult::costImpact).sum(),
				true, h, ph);
	}

	private static List<IPlanRewriteRule> toList(Iterable<IPlanRewriteRule> rules) {
		var r = new ArrayList<IPlanRewriteRule>();
		for (var x : rules)
			r.add(x);
		return r;
	}

	private static List<IPlanRewriteRule> orderRules(List<IPlanRewriteRule> rules) {
		var byName = new HashMap<String, IPlanRewriteRule>();
		for (var r : rules)
			if (byName.put(r.name(), r) != null)
				throw new IllegalArgumentException("Duplicate rewrite rule '" + r.name() + "'.");
		for (var r : rules) {
			for (var dep : r.mustRunAfter())
				if (!byName.containsKey(dep))
					throw new IllegalStateException(
							"Rewrite rule '" + r.name() + "' references unknown rule '" + dep + "'.");
			for (var dep : r.mustRunBefore())
				if (!byName.containsKey(dep))
					throw new IllegalStateException(
							"Rewrite rule '" + r.name() + "' references unknown rule '" + dep + "'.");
			for (var conflict : r.conflictsWith())
				if (byName.containsKey(conflict) && byName.get(conflict).conflictsWith().contains(r.name()))
					throw new IllegalStateException(
							"Rewrite rules '" + r.name() + "' and '" + conflict + "' conflict and cannot be composed.");
		}
		var edges = new HashMap<String, Set<String>>();
		var indegree = new HashMap<String, Integer>();
		for (var r : rules) {
			edges.put(r.name(), new HashSet<>());
			indegree.put(r.name(), 0);
		}
		for (var r : rules) {
			for (var after : r.mustRunAfter())
				addEdge(edges, indegree, after, r.name());
			for (var before : r.mustRunBefore())
				addEdge(edges, indegree, r.name(), before);
		}
		var queue = new PriorityQueue<IPlanRewriteRule>(
				Comparator.comparingInt(IPlanRewriteRule::priority).thenComparing(IPlanRewriteRule::name));
		for (var r : rules)
			if (indegree.get(r.name()) == 0)
				queue.add(r);
		var ordered = new ArrayList<IPlanRewriteRule>();
		while (!queue.isEmpty()) {
			var r = queue.remove();
			ordered.add(r);
			for (var next : edges.get(r.name()))
				if (indegree.merge(next, -1, Integer::sum) == 0)
					queue.add(byName.get(next));
		}
		if (ordered.size() != rules.size())
			throw new IllegalStateException("Rewrite rule composition contains an ordering cycle.");
		return List.copyOf(ordered);
	}

	private static void addEdge(Map<String, Set<String>> edges, Map<String, Integer> indegree, String from, String to) {
		if (edges.get(from).add(to))
			indegree.merge(to, 1, Integer::sum);
	}

	public static PlanRewriteRuleResult applyRule(IPlanRewriteRule rule, SemanticPlan before, SemanticPlan after) {
		var security = SecurityPreservationProof.create(before, after);
		var obligation = SecurityObligationProof.create(rule, before, after);
		var semantic = SemanticEquivalenceProof.create(before, after);
		var authorization = SemanticPlanAuthorizationBindingProof.create(before, after);
		var optimization = PlanOptimizationProof.create(rule, before, after);
		return new PlanRewriteRuleResult(rule.name(), before, after, rule.preconditions(), rule.securityObligations(),
				rule.costImpact(), security, obligation, semantic, authorization, optimization);
	}
}
